app/src/main/java/science/apolline/service/sensor/BluetoothLeService.kt

```diff
@@ -0,0 +1,620 @@
+                    broadcastUpdate(intentAction)
+                    Log.i(TAG, "Connected to GATT server.")
+                    // Attempts to discover services after successful connection.
+                    Log.i(TAG, "Attempting to start service discovery:" + mBluetoothGatt!!.discoverServices())
+
+                    val UIHandl = Handler(Looper.getMainLooper())
+                    UIHandl.postDelayed({
+                        Toast.makeText(this@BluetoothLeService, "Start", Toast.LENGTH_SHORT).show()
+                        if (MainActivity.mNotifyCharacteristic != null) {
+                            MainActivity.mNotifyCharacteristic!!.setValue("c")
+                            MainActivity.mBluetoothLeService!!.writeCharacteristic(MainActivity.mNotifyCharacteristic!!)
+                        }
+                    }, 5000)
+
+                } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
+                    intentAction = ACTION_GATT_DISCONNECTED
+                    Log.i(TAG, "Disconnected from GATT server.")
+                    broadcastUpdate(intentAction)
+                }
+            }
+
+            override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
+                if (status == BluetoothGatt.GATT_SUCCESS) {
+                    broadcastUpdate(ACTION_GATT_SERVICES_DISCOVERED)
+                    println("hihihihihi 3" )
+                } else {
+                    Log.w(TAG, "onServicesDiscovered received: $status")
+                }
+            }
+
+            override fun onCharacteristicRead(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic, status: Int) {
+                if (status == BluetoothGatt.GATT_SUCCESS) {
+                    broadcastUpdate(ACTION_DATA_AVAILABLE, characteristic)
+                    println("hihihihihi 1" )
+                }
+            }
+
+            @SuppressLint("SetTextI18n")
+            override fun onCharacteristicChanged(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic) {
+
+
+
+                val tempBuff = characteristic.getStringValue(0)
+                buff += tempBuff
+
+                println("hihihihihi " + buff )
+
+
+
+                if (buff.contains("\n")) {
+                    if (!buff.contains("#") && buff.indexOf("2") == 0) {
+                        //Decode rawdata
+                        //date
+                        appaData.dateGPS = buff.substring(0, buff.indexOf(";"))
+                        //PM1
+                        var debut = buff.indexOf(";")
+                        var fin = buff.indexOf(";", debut + 1)
+                        var Value = buff.substring(debut + 1, fin)
+                        try {
+                            appaData.pm01Value = Integer.parseInt(Value)
+                        } catch (e: NumberFormatException) {
+                            appaData.pm01Value = 0
+                        }
+
+                        //PM2.5
```

Comment:
looks like duplicate code. Please apply DRY principle and extract as a private function.